### DISCUSSION TITLE
How can I help?

### DISCUSSION THREAD

**Discussion by peternrdstrm** [bromite/bromite#2066] [General]
I've been roaming this github page for a long time now, so I think it's time to intoduce myself. 
I'm Peter, a highschool maths&computer science major obsessed with privacy and security, among other thing such as electronic music and indie videogames. 
I've been using Bromite for almost two years now, and I really, I mean really love it now. 
I also really look up to both @csagan5 and @uazo 's dedication to the project. 
As for my skills, I'm starting to grow more and more comfortable with programming, dominantly in the Rust and C++ languages. I've also done web development in the past, but never took that very far. 
I'm also very familiar with android as a platform, and general thinkering with apps. 
So anyways, is there something I can help with in this project? I'd really like to support you guys

**Comment by csagan5** [comment]
> I've been using Bromite for almost two years now, and I really, I mean really love it now.

I started this project because I was not happy about the available mobile browser choices; at some point in 2017 I released it so that others could benefit from it too and hopefully contribute.
All Bromite patches are open source (GPLv3), which is just one of many ways to keep the code open and make a statement about the absence of any for-profit mechanisms in the browser.



> I also really look up to both @csagan5 and @uazo 's dedication to the project.
> As for my skills, I'm starting to grow more and more comfortable with programming, dominantly in the Rust and C++ languages. I've also done web development in the past, but never took that very far.

@uazo has contributed massive work for some time now and he has improved himself his skills (both in&outside the Bromite project, I presume) over time, so you could also see this as an opportunity to improve your skills in such sense.
Another similar example is @wchen342 (who now works for Brave IIRC) and started [ungoogled-chromium for Android](https://github.com/ungoogled-software/ungoogled-chromium-android)


> I'm also very familiar with android as a platform, and general tinkering with apps.

The Android platform is continuously evolving so that familiarity and attitude will surely help if you want to hang around here.

> So anyways, is there something I can help with in this project? I'd really like to support you guys

First and foremost: you need to be cautious with open source contributions because they can be very rewarding but also very taxing. You can expect all sort of toxic behaviours online and this place is no different than any other in this sense (despite the moderation I sometimes try to do). On this topic, as a general rule for the future I invite you to protect yourself by not engaging in any conversation which is not respectful towards you or others; I will stop such behaviours as soon as I am aware.

Having said that: any type of technical discussion (and the patch that sometimes follows) is very welcome. Bromite currently has around 200 patches with focus on:
1. privacy
2. security
3. UI changes which allow for better user control (often related to privacy, security or data ownership)
4. other unrelated UI patches (these are exceptions)

The goal is to stay on the latest Chromium major release however that is not guaranteed: this is a volunteers project, for example if you had to choose between working on Bromite or some leisure time or spending time with your loved ones, you should always not work on Bromite.

# How is the development work organised?
It is currently loosely organised; I have tried in the past to organise everything a bit better but the intent fails for a reason or another. For the time being I am happy to keep it loosely organised as I see it as a natural balance.

The bulk of work can be categorised as:
* maintenance of existing patches
* new patches/bugfixes
* development tooling improvements
* open source promotion

Sadly this is a list sorted from "most developed" to "least developed", as I have not often had time to attract more contributors, but I am always happy if someone wants to join. Also regarding "development tooling improvements": @uazo has developed a few interesting tools already, however I have not officially accepted any as I am working on another set of tools which I am evaluating when/how to release. The tools cover different aspects of the development cycle so they would not overlap anyways.

# How to contribute

I would suggest reading the existing open issues looking for something you might like to work on; there are also forgotten/lost ideas in the issue tracker that were not rejected but not accepted either for the lack of contributors, see [this filter](https://github.com/bromite/bromite/issues?q=is%3Aissue+label%3Aenhancement-without-maintainer+is%3Aclosed).
Finally you can also contribute your own ideas but it is advised to explore them a bit before submissions: it might have already been discussed in some other issue or discussion.

In parallel to this activity I suggest that you start building Chromium for Android, that will bring you up to speed on a lot of aspects of Chromium-based browsers. From time to time we discuss with @uazo about new Chromium features and what to do in Bromite, so you can as well participate there ([example](https://github.com/bromite/bromite/issues/2052)) and (when you feel ready) volunteer to submit some patch.

**Reply by peternrdstrm** [reply]
Thank you, I really apreciate your kindness. 
I think I will try to find a simple patch to begin with, something really minor maybe. Going forward, I do have big ideas to work on, but with no prior experience in chromium development I can't say if I'll be able to get them done. I'll also have to get comfortable with Java probably, but I've already written some minor projects in Kotlin. 
Either way, I might begin this weekend, if I'll have the time

**Reply by wchen342** [reply]
Adding my 2 cents, maybe also for future references:

1. Don't be afraid to mess around! Chromium source codes may look complicated, but you can always start from changing something small, like a variable, and observe its effects. Also be sure to use logging tools, e.g. `LOG(WARNING)` for C++ and `Log.w()` for Java, which are very helpful when debugging.
2. Some good starting point will be looking into [how to get chromium source and start building](https://chromium.googlesource.com/chromium/src/+/main/docs/android_build_instructions.md), [what `gn` files are](https://chromium.googlesource.com/chromium/src/tools/gn/+/48062805e19b4697c5fbd926dc649c78b6aaa138/README.md), and [chromium source folder structures](https://www.chromium.org/developers/how-tos/getting-around-the-chrome-source-code/).
3. Set an adequate goal. Maintaining a chromium fork is hard and very time consuming, since chromium updates very frequently and you are looking at a project that is maintained by hundreds of people. Try to keep up and take things step by step. I used to work overtime all the time, and that took its toll on me eventually. Do not do that.
4. I will suggest a machine with 16+GB memory, a CPU with a lot of cores (more cores is more important), 500+GB SSD and a fast internet connection for syncing chromium source (>1MB/s).
5. Seek help when needed. I preferred to work on my own but in retrospect I probably should have discussed more with others. The developers from Bromite and the main repo of ungoogled-chromium are mostly helpful.
